src/integrators/batchNormNN.jl

```diff
@@ -0,0 +1,190 @@
+        if this.outTimes[i]==1
+            Ydata = [Ydata; this.Q*Yd]
+        end
+        if doDerivative
+            tmp[i+1,1] = copy(Y)
+        end
+        cnt = cnt + ni
+    end
+
+    return Ydata,Y,tmp
+end
+
+# -------- Jacobian matvecs --------
+function JYmv(this::batchNormNN{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+    nex = div(length(Y),nFeatIn(this))
+    nt = length(this.layers)
+    cnt = 0
+    dYdata = zeros(T,0,nex)
+    for i=1:nt
+        ni = nTheta(this.layers[i])
+        dY = JYmv(this.layers[i],dY,theta[cnt+(1:ni)],tmp[i,1],tmp[i,2])[2]
+        if this.outTimes[i]==1
+            dYdata = [dYdata; this.Q*dY]
+        end
+        cnt = cnt+ni
+    end
+    return dYdata, dY
+end
+
+function  Jmv(this::batchNormNN{T},dtheta::Array{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+    nex = div(length(Y),nFeatIn(this))
+    nt = length(this.layers);
+    if isempty(dY)
+        dY = 0*Y
+    end
+
+    dYdata = zeros(T,0,nex)
+    cnt = 0
+    for i=1:nt
+        ni = nTheta(this.layers[i])
+        dY = Jmv(this.layers[i],dtheta[cnt+(1:ni)],dY,theta[cnt+(1:ni)],
+                tmp[i,1],tmp[i,2])[2]
+        if this.outTimes[i]==1
+            dYdata = [dYdata; this.Q*dY]
+        end
+        cnt = cnt+ni
+    end
+    return dYdata,dY
+end
+
+# -------- Jacobian' matvecs --------
+function JYTmv(this::batchNormNN{T},Wdata::Array{T},W::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+
+    nex = div(length(Y),nFeatIn(this));
+    if !isempty(Wdata)
+        Wdata = reshape(Wdata,:,nex);
+    end
+    if isempty(W)
+        W = zero(T)
+    elseif length(W)>1
+        W     = reshape(W,:,nex)
+    end
+    nt = length(this.layers)
+
+    cnt = 0; cnt2 = 0;
+    for i=nt:-1:1
+        ni = nTheta(this.layers[i])
+        if this.outTimes[i]==1
+            nn = nFeatOut(this.layers[i])
+            W = W + this.Q'*Wdata[end-cnt2-nn+1:end-cnt2,:]
+            cnt2 = cnt2 + nn
+        end
+        W  = JYTmv(this.layers[i], W,(T)[],theta[end-cnt-ni+1:end-cnt],
+                    tmp[i,1],tmp[i,2])
+        cnt = cnt+ni
+    end
+    return W
+end
+
+
+function JthetaTmv(this::batchNormNN{T},Wdata::Array{T},W::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+	return JTmv(this,Wdata,W,theta,Y,tmp)[1];
+end
+
+
+
+function JTmv(this::batchNormNN,Wdata::Array{T},W::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+    
+    nex = div(length(Y),nFeatIn(this))
+
+    if size(Wdata,1)>0
+        Wdata = reshape(Wdata,:,nex)
+    end
+    if length(W)==0
+        W = zeros(T,nFeatOut(this),nex)
+    elseif length(W)>1
+        W     = reshape(W,:,nex)
+    end
+
+    dtheta = 0*theta
+    nt = length(this.layers)
+
+    cnt = 0; cnt2 = 0
+    for i=nt:-1:1
+        if this.outTimes[i]==1
+            nn = nFeatOut(this.layers[i])
+            W += this.Q'*Wdata[end-cnt2-nn+1:end-cnt2,:]
```

Comment:
Depending on how big this sub array is, it might be worth using a @view to avoid the allocation

Reply:
These are all things that are also in NN so lets put this on the TODO list and make the changes together